Add create_game_event template

Danielsio opened the pull request:
### Desired Outcome

1. users will have a template to create game events
2. the template will have routing like "/game-events/create"

### Implemented Changes
1. I added a template for the form to create game events.
2. added a view that will render the template + will accept the post reqeust from the user and create the game event
3. added basic validation  for the view
4. applied group routing for game-events 

Most of the PR is the game_events.views.create_game_event function
and the template itself game_event\templates\game_event\create_game_event.html

### Connected Issue/Story

Resolves #68 

### Dependencies

There are no dependencies for now.
- [x] This PR does not depend on other PR's 

### Definition of Done

1. Form for creation of GameEvents for the Players will be displayed at routing **`"game-events/create"`**
2. A view will be created which handles the display of the form and the creation GameEvent from the filled form by the Player.
3. Form Data will be validated inside the **`GameEvent.create()`** function, and if valid will create the GameEvent. 
4. Otherwise error messeages will be displayed and the player will be **`redirected back to the form page`**.

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code changes

#### Documentation
- [x] This PR does not require updating any documentation

amitCohen2 commented on meetballs/urls.py:
```diff
@@ -14,11 +14,12 @@
     path('admin/', admin.site.urls),
-    path('', views.home, name='home'),
+    path('', app_views.home, name='home'),
+    path('game-events/', include("game_event.urls"))
```
May be better to add at the end of this line ','  to avoid problems later.
Should look like that this line:  path('game-events/', include("game_event.urls")),

Danielsio replied:
Sure

Danielsio commented on game_event/views.py:
```diff
@@ -1 +1,56 @@
+        })
+
+
+def validate_game_event_form(request, court, ball_game, min_number_of_players, time):
```
i am not sure how to test this to be honest.. @yftacherzog @Yuval-Vino maybe you know simple way to verify that form logic is ok ?

Yuval-Vino replied:
In my opinion, you can move all validators to the create function of the class (which is missing btw). Then, in the `create_game_event `function, you could try to create the game event and catch any errors that occur.

After that, you could add some tests to the create function to ensure that it works correctly.

For the views tests, you could send a valid request and check that the game event is created, and also send an invalid request and make sure that the game event hasn't been created.

Danielsio replied:
I managed to get it to work somehow

Yuval-Vino commented on game_event/models.py:
```diff
@@ -17,3 +20,45 @@ class GameEvent(models.Model):
     )
+
+    @staticmethod
+    def create_game_event_from_form(request):
```
My intention was to create a function (similar to the one in #47) that incorporates validators.
This way, from the views, you can simply call the create method(which shouldnt require a request input, but only relevant fields for `GameEvent`). If an error occurs during the creation process, it can be caught in the views and a relevant error message can be sent

This way the creation of `GameEvent` will be more robust, and not force it to be in a request form

Danielsio replied:
i fixed it , but in this current approch only one messeage will be displayed to the user.. the first ValidationError.

Yuval-Vino replied:
I suggest modifying the function to throw an exception with a string instead of a list. This will make it more user-friendly for future users who may expect a string output.
To achieve this, you can init errors to an empty string and concatenating relevant errors using `errors += 'relevant error'` instead of repeatedly appending to the list. 
Alternatively, if you want to keep the current Implementation you can use the join method to concatenate the errors list  
  `" ".join(errors)`.

What do you think?

Danielsio replied:
could you elaborate ? I didnt understand the difference in your approch.

Danielsio replied:
It's a little uncomfortable this way, cause i want to later iterate over the errors and print messeage to the user

Yuval-Vino commented on game_event/tests.py:
```diff
@@ -75,3 +82,32 @@ def test_create_game_event_with_null_ball_game(self, saved_court):
+            'ball_game': TEST_BALL_GAME,
+        }
+        client.post(url, data=form_data)
+        assert GameEvent.objects.count() == 1
```
This test will fail when you have data. To prevent this, consider adding the previous size of the data as a variable before running the test. Then, you can test that the new size is equal to the previous size plus one

Danielsio replied:
Fixed

Yuval-Vino commented on game_event/tests.py:
```diff
@@ -75,3 +82,32 @@ def test_create_game_event_with_null_ball_game(self, saved_court):
+            'ball_game': TEST_BALL_GAME,
+        }
+        client.post(url, data=data)
+        assert GameEvent.objects.count() == 0
```
same

Danielsio replied:
Fixed

Yuval-Vino commented on court_ball_game/models.py:
```diff
@@ -12,3 +12,11 @@ class CourtBallGame(models.Model):
+    @staticmethod
+    def is_ball_game_playable(court, ball_game):
+        ball_games_available_at_court = CourtBallGame.objects.filter(court=court)
+        for bg in ball_games_available_at_court:
+            if bg.ball_game == ball_game:
+                return True
+        return False
```
More pythonic
```suggestion
    return any(bg.ball_game == ball_game for bg in ball_games_available_at_court)

```

Danielsio replied:
Thanks

Danielsio commented on game_event/urls.py:
```diff
@@ -0,0 +1,7 @@
+
+
+urlpatterns = [
+    path('create/', views.create_game_event, name='create game event'),
```
Thanks, That is a really good advice, will be changed to display_game_event_form.

Yuval-Vino commented on game_event/models.py:
```diff
@@ -17,3 +20,27 @@ class GameEvent(models.Model):
+        return game_event
+
+    @staticmethod
+    def validate_game_event(court, ball_game, min_number_of_players, time):
```
why dont we validate `max_number_of_players`?

Danielsio replied:
you are right, i can validate if max - min is > 0

MaayanMashhadi commented on game_event/models.py:
```diff
@@ -17,3 +20,27 @@ class GameEvent(models.Model):
+        if min_number_of_players < 1:
+            errors.append("The minimum number of players must be at least 1.")
+        if time < timezone.now():
+            errors.append("The selected time must be in the future.")
```
Maybe consider to add upper bound about the time of the event.
I think maybe to add event just in the current month. (Cause I think to start some event in the far future, like more than half a year isn't so good, it could mess up the web). 
Would like to hear your opinion about it.

Danielsio replied:
i will change it to a month interval

MaayanMashhadi replied:
> i will change it to a month interval

I'm not sure you understand me correct so just for ensuring.
I mean to add another validate about the future (not the past), like something like this:
`if time.month > timezone.now().month+1: error`
this validate that the event will be in the current month or month after and not be held about 6 month after...

MaayanMashhadi commented:
Hi Daniel,
You did a good job here!
I left you some reviews, would like to hear what you think about it.

Danielsio commented on game_event/models.py:
```diff
@@ -17,3 +20,27 @@ class GameEvent(models.Model):
+        errors = []
+        if not CourtBallGame.is_ball_game_playable(court, ball_game):
+            errors.append("The selected court doesn't support the selected ball game.")
+        if min_number_of_players < 1:
```
sounds good

Danielsio commented:
![image](https://github.com/redhat-beyond/MeetBalls/assets/102610047/25c29d96-dba7-48eb-9924-1db806cca82e)

here are some screenshots of the form and the possible validation erros.

![image](https://github.com/redhat-beyond/MeetBalls/assets/102610047/cc0f364e-8e67-49cf-92cc-6115a07bef6d)

And here is just noraml form

Danielsio commented:
Dear Reviewers. 
It's Important that you let me know, if the current implementation of the value of the court is sufficient (for future perposes maybe), since it's only the concatenation of the city and the neighborhood, and I thought it would be more user friendly fot the creator of an event to understand what court he is choosing .

ZvikaNaorCohen commented:
In my opinion, we should keep the court value like that.

ZvikaNaorCohen commented on game_event/tests.py:
```diff
@@ -75,3 +84,32 @@ def test_create_game_event_with_null_ball_game(self, saved_court):
+            'level_of_game': 5,
+            'min_number_of_players': -1,
+            'max_number_of_players': 4,
+            'court': saved_court.courtID,
```
I think that `GameEvent` entity has a data member of court (`court = models.ForeignKey(Court, on_delete=models.CASCADE)`) and not `court.courtID`. 
It raised a few errors to me when I was dealing with creating new GameEvent entity. 
If we remove the .courtID and keep just the saved_court, will the tests fail?

Danielsio replied:
i belive you are right

Danielsio replied:
i get Field 'courtID' expected a number but got 'Court object (1)'. when i tried to send him only the saved_court

ZvikaNaorCohen commented:
> ![image](https://user-images.githubusercontent.com/102610047/237457092-25c29d96-dba7-48eb-9924-1db806cca82e.png)
> 
> here are some screenshots of the form and the possible validation erros.
> 
> ![image](https://user-images.githubusercontent.com/102610047/237457362-cc0f364e-8e67-49cf-92cc-6115a07bef6d.png)
> 
> And here is just noraml form

The validation tests look very good with that red warning. Well done.
What happens when a user wants to enter a court? Does he have a list of all the existing ones, and he picks from there?
Or does he write his own text, and a court would be created?

Danielsio commented on game_event/models.py:
```diff
@@ -17,3 +21,29 @@ class GameEvent(models.Model):
+        if min_number_of_players < 2:
+            errors.append("The minimum number of players must be at least 2.")
+        if max_number_of_players - min_number_of_players < 0:
+            errors.append("The Maximum number of players must be at least equal to the minimum number of players")
```
thanks, nice catch

Danielsio commented:
> The validation tests look very good with that red warning. Well done. What happens when a user wants to enter a court? Does he have a list of all the existing ones, and he picks from there? Or does he write his own text, and a court would be created?

For now, i render select option for each instance of a scourt, so user chooses from the existing courts in db.
maybe in the future we could manage a field for the user to input some address e.g "Kriat Ono Shlomo hameleh" and with the help of th "pin" library we can extract longitude and latitude from it, to create fresh instance of court in db. just a thought

MaayanMashhadi commented:
> Dear Reviewers. It's Important that you let me know, if the current implementation of the value of the court is sufficient (for future perposes maybe), since it's only the concatenation of the city and the neighborhood, and I thought it would be more user friendly fot the creator of an event to understand what court he is choosing .

For me it sounds good, I agree it will be more user friendly.

Danielsio commented:
Do you still need your routing at game_event.urls ? 

i dont know which PR will be merged first, so in a case this one get's merged, i will have to do quick rebase to include the game_event.urls in the meat_balls_urls

so i wonder maybe you can just fix it now  ? you can look at my PR #69 

and instead of the 'game-event' routing you got there you could replace that line with 
`path('', views.game_events, name='game events')`

the empty string in the path is because it has the prefix of "game-events/" from the meat_balls.urls

yftacherzog commented on court_ball_game/models.py:
```diff
@@ -12,3 +12,8 @@ class CourtBallGame(models.Model):
         unique_together = ('court', 'ball_game')
+
+    @staticmethod
+    def is_ball_game_playable(court, ball_game):
```
missing tests for this one

Danielsio replied:
added test for this

yftacherzog commented on court/models.py:
```diff
@@ -15,3 +15,7 @@ def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
+
+    @staticmethod
+    def get_all_courts():
```
missing tests for this one

Danielsio replied:
added one

yftacherzog commented on game_event/models.py:
```diff
@@ -17,3 +21,29 @@ class GameEvent(models.Model):
+        errors = []
+        if not CourtBallGame.is_ball_game_playable(court, ball_game):
+            errors.append("The selected court doesn't support the selected ball game.")
+        if min_number_of_players < 2:
```
Can't this validation be part of the model?

Danielsio replied:
not sure i understood
i wrap the call with the GameEvent.is_ball_game_playable_at_court()
if it wasnt your intension then i would need elaboration on this part

yftacherzog commented on game_event/models.py:
```diff
@@ -17,3 +21,29 @@ class GameEvent(models.Model):
+        return game_event
+
+    @staticmethod
+    def validate_game_event(court, ball_game, min_number_of_players, max_number_of_players, time):
```
missing tests

Danielsio replied:
Added tests

yftacherzog commented on game_event/models.py:
```diff
@@ -17,3 +21,29 @@ class GameEvent(models.Model):
     )
+
+    @staticmethod
+    def create(time, level_of_game, min_number_of_players, max_number_of_players, court, ball_game):
```
missing tests

yftacherzog commented on game_event/views.py:
```diff
@@ -1 +1,38 @@
+from django.contrib import messages
+
+
+def display_game_event_form(request):
```
missing tests.

yftacherzog replied:
Also, by having the same function used for both http methods, you're making this code harder to test. i.e. harder to maintain.

Danielsio replied:
took me some modifications, but i managed to split it to seperate views

Danielsio replied:
added test for display the form, and for post process

Danielsio commented on game_event/tests.py:
```diff
@@ -75,3 +84,32 @@ def test_create_game_event_with_null_ball_game(self, saved_court):
+        client.post(url, data=form_data)
+        assert GameEvent.objects.count() == count_of_game_events_before + 1
+
+    def test_create_game_event_invalid_form(self, client, saved_court):
```
I replaced this test with several validation tests

Danielsio commented on game_event/tests.py:
```diff
@@ -4,6 +4,11 @@
 from django.utils import timezone
 from player.models import BallGame
+from django.urls import reverse
+from .views import display_game_event_form  # import the view function
```
Nope

yftacherzog commented on court/tests.py:
```diff
@@ -127,3 +127,7 @@ def test_court_without_max_players(self):
+
+    def test_get_all_courts(self, court):
+        courts = Court.get_all_courts()
+        assert court in courts
```
You're testing that court is there rather than all courts. To test it in a robust way, you'd need to use the actual line that you use inside the method. That raises the question, do you really gain anything by adding this method? You could just replace any call to this custom method to a call to the built-in method that this custom method is calling. No? It will be just as easy to use and will mean less code to maintain.

Danielsio replied:
True, will be replaced

yftacherzog commented on court_ball_game/tests.py:
```diff
@@ -59,3 +59,6 @@ def test_create_court_ball_game_with_null_ball_game(self, court_instance):
                 court=court_instance
             ).full_clean()
+
+    def test_is_ball_game_playable(self, court_instance, court_ball_game_instance):
```
You proved here that it can return true. You need to also prove that it can return false (another test case or parametrize this test with pytest parametrize).

Danielsio replied:
added test with expected false

yftacherzog commented on game_event/views.py:
```diff
@@ -1 +1,42 @@
+
+        return redirect("/")  # will be changes later to "/game-events"
+    else:
+        redirect("/game-events/create")
```
What's the reason here it's `redirect` and earlier it's `return redirect`?

Danielsio replied:
my mistake

Danielsio commented:
@Yuval-Vino I hope I have addressed all your comments, you can update me if something remains unaddressed and I would be happy if you resolve the conversations.
